docs/src/literate/topology_optimization.jl

```diff
@@ -0,0 +1,502 @@
+# To store the density and the strain required to calculate the driving forces, we create the struct
+# `MaterialState`. We add a constructor to initialize the struct. The function `UpdateMaterialStates!`
+# updates the density values once we calculated the new values.
+
+mutable struct MaterialState{T, S <: AbstractArray{SymmetricTensor{2, 2, T}, 1}} 
+    χ::T ## density
+    ε::S ## strain in each quadrature point
+end
+
+function MaterialState(ρ, n_qp)
+    return MaterialState(ρ, Array{SymmetricTensor{2,2,Float64},1}(undef, n_qp))
+end
+
+function UpdateMaterialStates!(χn1, states, dh)
+    for (element, state) in zip(CellIterator(dh),states)
+        state.χ = χn1[cellid(element)]
+    end
+end
+#md nothing # hide
+
+# Next, we define a function to calculate the driving forces for all elements.
+# For this purpose, we iterate through all elements and calculate the average strain in each
+# element. Then, we compute the driving force from the formula introduced at the beginning.
+
+function DrivingForces(states, mp, dh)
+    pΨ = zeros(length(states))
+    χn = zeros(length(states))
+    for (element, state) in zip(CellIterator(dh), states)
+        i = cellid(element)
+        ε = sum(state.ε)/length(state.ε) ## average element strain
+        pΨ[i] = 1/2 * mp.p * state.χ^(mp.p-1) * (ε ⊡ mp.C ⊡ ε)
+        χn[i] = state.χ
+    end
+    return pΨ, χn 
+end
+#md nothing # hide
+
+# Now we calculate the Laplacian. For this purpose, we will later create the grid topology of 
+# the grid by using the function `ExclusiveTopology`. Then we iterate through each face of each element,
+# obtaining the neighboring element by using the `getneighborhood` function. For boundary faces,
+# the function call will return an empty object. In that case we use the dictionary to instead find the opposite
+# face, as discussed in the introduction. Then, the approximation of the Laplacian reduces to the sum below. 
+
+function Laplacian(dh, topology, χn, Δh)
+    ∇²χ = zeros(getncells(dh.grid))
+    _nfaces = nfaces(dh.grid.cells[1])
+    opp = Dict(1=>3, 2=>4, 3=>1, 4=>2)
+    nbg = zeros(Int,_nfaces)
+    
+    for element in CellIterator(dh)
+        i = cellid(element)
+        for j in 1:_nfaces
+            nbg_cellid = getcells(getneighborhood(topology, dh.grid, FaceIndex(i,j)))
+            if(!isempty(nbg_cellid))
+                nbg[j] = first(nbg_cellid) ## assuming only one face neighbor per cell
+            else ## boundary face
+                nbg[j] = first(getcells(getneighborhood(topology, dh.grid, FaceIndex(i,opp[j]))))
+            end
+        end
+        
+        ∇²χ[i] = (χn[nbg[1]]+χn[nbg[2]]+χn[nbg[3]]+χn[nbg[4]]-4*χn[i])/(Δh^2)
+    end
+
+    return ∇²χ
+end
+#md nothing # hide
+
+# To ensure that the mass is kept constant, we have to calculate the constraint
+# parameter $\lambda$, which we do via the bisection method. We repeat the calculation
+# until the difference between the current density and target density nearly vanishes. 
```

Comment:
```suggestion
# For the iterative computation of the solution, a function is needed to update the local densities.
# To ensure that the total mass is kept constant, we have to calculate the constraint
# parameter $\lambda$, which we do via the bisection method. We repeat the calculation
# until the difference between the current global density and target density nearly vanishes. 
```

I added a sentence in the beginning to make clear, that the goal of the function is to compute the densities and that the constraint parameter is only needed to do the computation.
To me it would be clear now.
I also added a distinction between "local" and "global" density, so the reader can better understand what ρ in the function is and why it is used as criterion for the iteration.
However, I am not sure if that is the appropriate wording to make that distinction.
I just didn't come up with a better idea.

Reply:
```suggestion
# For the iterative computation of the solution, a function is needed to update the densities in each element.
# To ensure that the mass is kept constant, we have to calculate the constraint
# parameter $\lambda$, which we do via the bisection method. We repeat the calculation
# until the difference between the average density (calculated from the elementwise trial densities) and the target density nearly vanishes. 
```

Reply:
I think your suggestion is already quite good. For someone familiar with the topic the notion of local/global densities should make sene. I slightly changed your proposition to make it a bit easier for newcomers to the topic to understand what is going on.